retrievalmarket/types.go

```diff
@@ -244,31 +233,93 @@ func IsTerminalStatus(status DealStatus) bool {
 
 // Params are the parameters requested for a retrieval deal proposal
 type Params struct {
-	Selector                *cbg.Deferred // V1
+	Selector                CborGenCompatibleNode // V1
 	PieceCID                *cid.Cid
 	PricePerByte            abi.TokenAmount
 	PaymentInterval         uint64 // when to request payment
 	PaymentIntervalIncrease uint64
 	UnsealPrice             abi.TokenAmount
 }
 
+// paramsBindnodeOptions is the bindnode options required to convert custom
+// types used by the Param type
+var paramsBindnodeOptions = []bindnode.Option{
+	CborGenCompatibleNodeBindnodeOption,
+	TokenAmountBindnodeOption,
+}
+
 func (p Params) SelectorSpecified() bool {
-	return p.Selector != nil && !bytes.Equal(p.Selector.Raw, cbg.CborNull)
+	return !p.Selector.IsNull()
 }
 
 func (p Params) IntervalLowerBound(currentInterval uint64) uint64 {
 	intervalSize := p.PaymentInterval
 	var lowerBound uint64
 	var target uint64
-	for target < currentInterval {
+	for target <= currentInterval {
 		lowerBound = target
 		target += intervalSize
 		intervalSize += p.PaymentIntervalIncrease
 	}
 	return lowerBound
 }
 
-func (p Params) NextInterval(currentInterval uint64) uint64 {
+// OutstandingBalance produces the amount owed based on the deal params
+// for the given transfer state and funds received
+func (p Params) OutstandingBalance(fundsReceived abi.TokenAmount, sent uint64, inFinalization bool) big.Int {
+	// Check if the payment covers unsealing
+	if fundsReceived.LessThan(p.UnsealPrice) {
+		return big.Sub(p.UnsealPrice, fundsReceived)
+	}
+
+	// if unsealing funds are received and the retrieval is free, proceed
+	if p.PricePerByte.IsZero() {
+		return big.Zero()
+	}
+
+	// Calculate how much payment has been made for transferred data
+	transferPayment := big.Sub(fundsReceived, p.UnsealPrice)
+
+	// The provider sends data and the client sends payment for the data.
+	// The provider will send a limited amount of extra data before receiving
+	// payment. Given the current limit, check if the client has paid enough
+	// to unlock the next interval.
+	minimumBytesToPay := p.IntervalLowerBound(sent)
+	if inFinalization {
+		// for last payment, we need to get past zero
+		minimumBytesToPay = sent
+	}
```

Comment:
```suggestion
	minimumBytesToPay := sent // for last payment, we need to get past zero
	if !inFinalization {
		minimumBytesToPay = p.IntervalLowerBound(sent)
	}
```

mildly pedantic, but there's a loop in `IntervalLowerBound`